File: backend/app/providers/claude.py
"""Provedor Claude: OAuth local + contas extras coladas no painel."""

from __future__ import annotations

import time
from typing import Any

from app.formatting import as_percent_points, claude_utilization_percent, iso_or_none, pick
from app.http_util import http_json
from app.local.claude_oauth import claude_token_candidates
from app.store import provider as provider_cfg

CLAUDE_USAGE_URL = "https://api.anthropic.com/api/oauth/usage"
CLAUDE_BETA = "oauth-2025-04-20"
CLAUDE_USER_AGENT = "claude-code/2.1"

SCOPE_HINT = (
    "token sem escopo user:profile (típico de claude setup-token / oat). "
    "Apague o token colado no painel e use o Claude Code logado neste Mac (`claude` / /login)."
)


def claude_fail(msg: str) -> dict[str, Any]:
    return {
        "ok": False,
        "error": msg,
        "session_percent": None,
        "session_resets_at": None,
        "weekly_percent": None,
        "weekly_resets_at": None,
        "sonnet_percent": None,
        "sonnet_resets_at": None,
        "opus_percent": None,
        "opus_resets_at": None,
    }


def parse_claude_payload(data: dict[str, Any]) -> dict[str, Any]:
    def _win(obj: Any) -> tuple[float | None, str | None]:
        if not isinstance(obj, dict) or not obj:
            return None, None
        pct = pick(
            as_percent_points(obj.get("percent")),
            claude_utilization_percent(obj.get("utilization")),
        )
        return pct, iso_or_none(pick(obj.get("resets_at"), obj.get("resetsAt")))

    session_pct, session_reset = _win(data.get("five_hour") or data.get("fiveHour"))
    weekly_pct, weekly_reset = _win(data.get("seven_day") or data.get("sevenDay"))
    sonnet_pct, sonnet_reset = _win(data.get("seven_day_sonnet") or data.get("sevenDaySonnet"))
    opus_pct, opus_reset = _win(data.get("seven_day_opus") or data.get("sevenDayOpus"))

    limits = data.get("limits")
    if isinstance(limits, list):
        for item in limits:
            if not isinstance(item, dict):
                continue
            kind = str(item.get("kind") or "").lower()
            pct = pick(
                as_percent_points(item.get("percent")),
                claude_utilization_percent(item.get("utilization")),
            )
            reset = iso_or_none(pick(item.get("resets_at"), item.get("resetsAt")))
            if kind in ("session", "five_hour", "5h") and session_pct is None:
                session_pct, session_reset = pct, reset
            if kind in ("weekly_all", "seven_day", "weekly", "7d") and weekly_pct is None:
                weekly_pct, weekly_reset = pct, reset
            if "sonnet" in kind and sonnet_pct is None:
                sonnet_pct, sonnet_reset = pct, reset
            if "opus" in kind and opus_pct is None:
                opus_pct, opus_reset = pct, reset

    ok = any(v is not None for v in (session_pct, weekly_pct, sonnet_pct, opus_pct))
    return {
        "ok": ok,
        "error": None if ok else "resposta Claude sem janelas de cota",
        "session_percent": session_pct,
        "session_resets_at": session_reset,
        "weekly_percent": weekly_pct,
        "weekly_resets_at": weekly_reset,
        "sonnet_percent": sonnet_pct,
        "sonnet_resets_at": sonnet_reset,
        "opus_percent": opus_pct,
        "opus_resets_at": opus_reset,
    }


def _is_scope_error(msg: str) -> bool:
    low = msg.lower()
    return "user:profile" in low or "oauth_scope_insufficient" in low


def fetch_claude_one(token: str) -> dict[str, Any]:
    token = (token or "").strip()
    if not token:
        return claude_fail("sem token Claude")
    try:
        data = http_json(
            CLAUDE_USAGE_URL,
            headers={
                "Authorization": f"Bearer {token}",
                "anthropic-beta": CLAUDE_BETA,
                "Accept": "application/json",
                "User-Agent": CLAUDE_USER_AGENT,
            },
            provider="CLAUDE",
        )
    except RuntimeError as exc:
        msg = str(exc)
        return claude_fail(SCOPE_HINT if _is_scope_error(msg) else msg)
    if not isinstance(data, dict):
        return claude_fail("resposta Claude inesperada")
    parsed = parse_claude_payload(data)
    if parsed.get("ok"):
        return parsed
    return claude_fail(str(parsed.get("error") or "resposta Claude sem janelas de cota"))


def _fetch_claude_local(cands: list[tuple[str, str, int | None]]) -> dict[str, Any]:
    last_err = "sem token Claude"
    now_ms = int(time.time() * 1000)
    for _source, token, exp_ms in cands:
        if exp_ms and exp_ms < now_ms:
            last_err = "OAuth expirado; abra o Claude Code neste Mac"
            continue
        result = fetch_claude_one(token)
        if result.get("ok"):
            return result
        last_err = str(result.get("error") or last_err)
    return claude_fail(last_err)


def fetch_claude_accounts(cfg: dict) -> list[dict[str, Any]]:
    accounts: list[dict[str, Any]] = []
    p = provider_cfg(cfg, "claude")

    if not p.get("hidden"):
        local_cands = claude_token_candidates(cfg)
        if local_cands:
            result = _fetch_claude_local(local_cands)
            accounts.append({"id": "local", "label": str(p.get("local_label") or ""), **result})

    extra = list(p.get("accounts") or [])
    if not extra:
        legacy = str(p.get("paste_secret") or "").strip()
        if legacy:
            extra = [{"id": "legacy", "label": "", "secret": legacy}]
    for acc in extra:
        token = str(acc.get("secret") or "").strip()
        label = str(acc.get("label") or "").strip()
        aid = str(acc.get("id") or "extra")
        result = fetch_claude_one(token)
        accounts.append({"id": aid, "label": label, **result})
    return accounts

File: backend/src/providers/claude.ts
/**
 * Provedor Claude: OAuth local + contas extras coladas no painel.
 * Port of backend-python-legacy/app/providers/claude.py
 */
import { asPercentPoints, claudeUtilizationPercent, isoOrNone, pick } from "../formatting.js";
import { httpJson } from "../httpClient.js";
import { claudeTokenCandidates } from "../local/claudeOauth.js";
import { provider as providerCfg } from "../store.js";

export const CLAUDE_USAGE_URL = "https://api.anthropic.com/api/oauth/usage";
export const CLAUDE_BETA = "oauth-2025-04-20";
export const CLAUDE_USER_AGENT = "claude-code/2.1";

export const SCOPE_HINT =
  "token sem escopo user:profile (típico de claude setup-token / oat). " +
  "Apague o token colado no painel e use o Claude Code logado neste computador (`claude` / /login).";

export function claudeFail(msg: string): Record<string, unknown> {
  return {
    ok: false,
    error: msg,
    session_percent: null,
    session_resets_at: null,
    weekly_percent: null,
    weekly_resets_at: null,
    sonnet_percent: null,
    sonnet_resets_at: null,
    opus_percent: null,
    opus_resets_at: null,
  };
}

export function parseClaudePayload(data: Record<string, unknown>): Record<string, unknown> {
  function win(obj: unknown): [number | null, string | null] {
    if (obj === null || typeof obj !== "object" || Array.isArray(obj) || Object.keys(obj as object).length === 0) return [null, null];
    const d = obj as Record<string, unknown>;
    const pct = pick(asPercentPoints(d.percent), claudeUtilizationPercent(d.utilization)) as number | null;
    return [pct, isoOrNone(pick(d.resets_at, d.resetsAt))];
  }

  let [sessionPct, sessionReset] = win((data.five_hour ?? data.fiveHour) as unknown);
  let [weeklyPct, weeklyReset] = win((data.seven_day ?? data.sevenDay) as unknown);
  let [sonnetPct, sonnetReset] = win((data.seven_day_sonnet ?? data.sevenDaySonnet) as unknown);
  let [opusPct, opusReset] = win((data.seven_day_opus ?? data.sevenDayOpus) as unknown);

  const limits = data.limits;
  if (Array.isArray(limits)) {
    for (const item of limits) {
      if (item === null || typeof item !== "object" || Array.isArray(item)) continue;
      const it = item as Record<string, unknown>;
      const kind = String(it.kind ?? "").toLowerCase();
      const pct = pick(asPercentPoints(it.percent), claudeUtilizationPercent(it.utilization)) as number | null;
      const reset = isoOrNone(pick(it.resets_at, it.resetsAt));
      if ((kind === "session" || kind === "five_hour" || kind === "5h") && sessionPct === null) {
        sessionPct = pct;
        sessionReset = reset;
      }
      if ((kind === "weekly_all" || kind === "seven_day" || kind === "weekly" || kind === "7d") && weeklyPct === null) {
        weeklyPct = pct;
        weeklyReset = reset;
      }
      if (kind.includes("sonnet") && sonnetPct === null) {
        sonnetPct = pct;
        sonnetReset = reset;
      }
      if (kind.includes("opus") && opusPct === null) {
        opusPct = pct;
        opusReset = reset;
      }
    }
  }

  const ok = [sessionPct, weeklyPct, sonnetPct, opusPct].some((v) => v !== null && v !== undefined);
  return {
    ok,
    error: ok ? null : "resposta Claude sem janelas de cota",
    session_percent: sessionPct,
    session_resets_at: sessionReset,
    weekly_percent: weeklyPct,
    weekly_resets_at: weeklyReset,
    sonnet_percent: sonnetPct,
    sonnet_resets_at: sonnetReset,
    opus_percent: opusPct,
    opus_resets_at: opusReset,
  };
}

function isScopeError(msg: string): boolean {
  const low = msg.toLowerCase();
  return low.includes("user:profile") || low.includes("oauth_scope_insufficient");
}

export async function fetchClaudeOne(token: string): Promise<Record<string, unknown>> {
  token = (token ?? "").trim();
  if (!token) return claudeFail("sem token Claude");
  let data: unknown;
  try {
    data = await httpJson(CLAUDE_USAGE_URL, {
      headers: {
        Authorization: `Bearer ${token}`,
        "anthropic-beta": CLAUDE_BETA,
        Accept: "application/json",
        "User-Agent": CLAUDE_USER_AGENT,
      },
      provider: "CLAUDE",
    });
  } catch (e) {
    const msg = String(e);
    return claudeFail(isScopeError(msg) ? SCOPE_HINT : msg);
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) return claudeFail("resposta Claude inesperada");
  const parsed = parseClaudePayload(data as Record<string, unknown>);
  if (parsed.ok) return parsed;
  return claudeFail(String(parsed.error ?? "resposta Claude sem janelas de cota"));
}

async function fetchClaudeLocal(cands: Array<[string, string, number | null]>): Promise<Record<string, unknown>> {
  let lastErr = "sem token Claude";
  const nowMs = Date.now();
  for (const [, token, expMs] of cands) {
    if (expMs !== null && expMs !== undefined && expMs < nowMs) {
      lastErr = "OAuth expirado; abra o Claude Code neste computador";
      continue;
    }
    const result = await fetchClaudeOne(token);
    if (result.ok) return result;
    lastErr = String(result.error ?? lastErr);
  }
  return claudeFail(lastErr);
}

export async function fetchClaudeAccounts(cfg: Record<string, unknown>): Promise<Array<Record<string, unknown>>> {
  const accounts: Array<Record<string, unknown>> = [];
  const p = providerCfg(cfg, "claude") as Record<string, unknown>;

  if (!p.hidden) {
    const localCands = claudeTokenCandidates(cfg);
    if (localCands.length > 0) {
      const result = await fetchClaudeLocal(localCands);
      accounts.push({ id: "local", label: String(p.local_label ?? ""), ...result });
    }
  }

  let extra = Array.isArray(p.accounts) ? [...(p.accounts as unknown[])] : [];
  if (extra.length === 0) {
    const legacy = String(p.paste_secret ?? "").trim();
    if (legacy) extra = [{ id: "legacy", label: "", secret: legacy }];
  }
  for (const accRaw of extra) {
    const acc = accRaw as Record<string, unknown>;
    const token = String(acc.secret ?? "").trim();
    const label = String(acc.label ?? "").trim();
    const aid = String(acc.id ?? "extra");
    const result = await fetchClaudeOne(token);
    accounts.push({ id: aid, label, ...result });
  }
  return accounts;
}

export const claude_fail = claudeFail;
export const parse_claude_payload = parseClaudePayload;
export const fetch_claude_one = fetchClaudeOne;
export const fetch_claude_accounts = fetchClaudeAccounts;
